File: backend/app/models/project.py
from __future__ import annotations

from sqlalchemy import JSON, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, PublishableMixin, SoftDeleteMixin, TimestampMixin
from app.models.tag import Tag, project_tag


class Project(TimestampMixin, PublishableMixin, SoftDeleteMixin, Base):
    __tablename__ = "project"

    id: Mapped[int] = mapped_column(primary_key=True)
    slug: Mapped[str] = mapped_column(String(140), unique=True, index=True, nullable=False)
    title_en: Mapped[str] = mapped_column(String(200), nullable=False)
    title_it: Mapped[str] = mapped_column(String(200), nullable=False)
    summary_en: Mapped[str] = mapped_column(Text, nullable=False)
    summary_it: Mapped[str] = mapped_column(Text, nullable=False)
    problem_en: Mapped[str] = mapped_column(Text, nullable=False)
    problem_it: Mapped[str] = mapped_column(Text, nullable=False)
    challenge_en: Mapped[str] = mapped_column(Text, nullable=False)
    challenge_it: Mapped[str] = mapped_column(Text, nullable=False)
    approach_en: Mapped[str] = mapped_column(Text, nullable=False)
    approach_it: Mapped[str] = mapped_column(Text, nullable=False)
    architecture_en: Mapped[str] = mapped_column(Text, nullable=False)
    architecture_it: Mapped[str] = mapped_column(Text, nullable=False)
    key_decisions_en: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    key_decisions_it: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    outcome_en: Mapped[str] = mapped_column(Text, nullable=False)
    outcome_it: Mapped[str] = mapped_column(Text, nullable=False)
    lessons_en: Mapped[str] = mapped_column(Text, nullable=False)
    lessons_it: Mapped[str] = mapped_column(Text, nullable=False)
    confidentiality_note_en: Mapped[str] = mapped_column(Text, nullable=False)
    confidentiality_note_it: Mapped[str] = mapped_column(Text, nullable=False)
    # Deprecated as the authoritative source (superseded by `tags`, shared
    # with Event via the same Tag vocabulary) but kept, unread by new code
    # paths once tags exist, purely so the one-time migration can normalize
    # existing data without a risky hard cutover — see event.py's `topics`.
    technologies: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    # Only rendered when present; never guessed.
    external_url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    is_featured: Mapped[bool] = mapped_column(default=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    # Admin-only — never serialized in any public schema.
    internal_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    skills: Mapped[list[ProjectSkill]] = relationship(
        back_populates="project", cascade="all, delete-orphan"
    )
    tags: Mapped[list[Tag]] = relationship(secondary=project_tag, order_by="Tag.label")


class ProjectSkill(Base):
    __tablename__ = "project_skill"

    id: Mapped[int] = mapped_column(primary_key=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("project.id"), nullable=False)
    skill_name: Mapped[str] = mapped_column(String(120), nullable=False)

    project: Mapped[Project] = relationship(back_populates="skills")

File: frontend/src/admin/types/project.ts
export interface AdminProjectListItem {
  id: number;
  slug: string;
  title_en: string;
  is_featured: boolean;
  sort_order: number;
  publication_status: "DRAFT" | "SCHEDULED" | "PUBLISHED" | "ARCHIVED";
  deleted_at: string | null;
}

export interface ProjectAdminDetail {
  id: number;
  slug: string;
  title_en: string;
  title_it: string;
  summary_en: string;
  summary_it: string;
  problem_en: string;
  problem_it: string;
  challenge_en: string;
  challenge_it: string;
  approach_en: string;
  approach_it: string;
  architecture_en: string;
  architecture_it: string;
  key_decisions_en: string[];
  key_decisions_it: string[];
  outcome_en: string;
  outcome_it: string;
  lessons_en: string;
  lessons_it: string;
  confidentiality_note_en: string;
  confidentiality_note_it: string;
  external_url: string | null;
  is_featured: boolean;
  sort_order: number;
  internal_notes: string | null;
  tags: string[];
  related_skills: string[];
  publication_status: "DRAFT" | "SCHEDULED" | "PUBLISHED" | "ARCHIVED";
  publish_at: string | null;
  unpublish_at: string | null;
  published_at: string | null;
  deleted_at: string | null;
  created_at: string;
  updated_at: string;
}

export type ProjectWritePayload = Omit<
  ProjectAdminDetail,
  | "id"
  | "slug"
  | "tags"
  | "related_skills"
  | "publication_status"
  | "publish_at"
  | "unpublish_at"
  | "published_at"
  | "deleted_at"
  | "created_at"
  | "updated_at"
> & { tag_labels: string[] };

export type ProjectCreatePayload = ProjectWritePayload & { slug: string };

export interface ProjectRevisionSummary {
  id: number;
  schema_version: number;
  snapshot: Record<string, unknown>;
  created_at: string;
  created_by_email: string | null;
}
